File: family-hub-complete/client/family_hub_client.py
"""Family Hub Python SDK — standard library only, no pip required."""
import json, urllib.request, urllib.error
from datetime import date
from typing import Optional, List

class FamilyHubError(Exception):
    def __init__(self,msg,status=0): super().__init__(msg); self.status=status

class FamilyHubClient:
    def __init__(self,base_url,api_key=None,timeout=10):
        self.base_url=base_url.rstrip('/'); self.api_key=api_key; self.timeout=timeout
        self.events=EventsResource(self); self.chores=ChoresResource(self)
        self.meals=MealsResource(self); self.grocery=GroceryResource(self); self.system=SystemResource(self)
    def _req(self,method,path,body=None):
        data=json.dumps(body).encode() if body else None
        headers={'Content-Type':'application/json','Accept':'application/json'}
        if self.api_key: headers['X-API-Key']=self.api_key
        req=urllib.request.Request(f"{self.base_url}{path}",data=data,headers=headers,method=method)
        try:
            with urllib.request.urlopen(req,timeout=self.timeout) as r: return json.loads(r.read())
        except urllib.error.HTTPError as e:
            try: d=json.loads(e.read())
            except: d={}
            raise FamilyHubError(d.get('error',f'HTTP {e.code}'),e.code)
        except urllib.error.URLError as e: raise FamilyHubError(str(e.reason))
    def get(self,p): return self._req('GET',p)
    def post(self,p,b=None): return self._req('POST',p,b)

class EventsResource:
    def __init__(self,c): self._c=c
    def list(self): return self._c.get('/api/events')
    def create(self,title,start,end=None,owner='Family',**kw): return self._c.post('/api/events',{'title':title,'start':start,'end':end,'owner':owner,**kw})
    def today(self): t=date.today().isoformat(); return [e for e in self.list() if (e.get('start') or '')[:10]==t]

class ChoresResource:
    def __init__(self,c): self._c=c
    def list(self): return self._c.get('/api/chores')
    def create(self,title,assignee='Family',due_date=None,points=5): return self._c.post('/api/chores',{'title':title,'assignee':assignee,'due_date':due_date or date.today().isoformat(),'points':points})
    def complete(self,chore_id): return self._c.post(f'/api/chores/{chore_id}/complete')
    def open(self): return [c for c in self.list() if c.get('status')=='open']

class MealsResource:
    def __init__(self,c): self._c=c
    def week(self): return self._c.get('/api/meals/week')
    def create(self,recipe,meal_type='dinner',meal_date=None,notes=''): return self._c.post('/api/meals',{'recipe':recipe,'meal_type':meal_type,'date':meal_date or date.today().isoformat(),'notes':notes})

class GroceryResource:
    def __init__(self,c): self._c=c
    def list(self): return self._c.get('/api/grocery/items')
    def add(self,name,quantity='1'): return self._c.post('/api/grocery/items',{'name':name,'quantity':quantity})
    def add_many(self,items): return [self.add(n,q) for n,q in items]

class SystemResource:
    def __init__(self,c): self._c=c
    def health(self): return self._c.get('/api/health')
    def is_online(self):
        try: self.health(); return True
        except: return False

def create_local_client(pi_ip='localhost',api_key=None): return FamilyHubClient(f'http://{pi_ip}:3001',api_key)
def create_remote_client(host,api_key): return FamilyHubClient(f'https://{host}',api_key,15)

if __name__=='__main__':
    import sys
    hub=create_local_client(sys.argv[1] if len(sys.argv)>1 else 'localhost', sys.argv[2] if len(sys.argv)>2 else None)
    print('Health:',hub.system.health())
    print('Today:',hub.events.today())
    print('Chores:',hub.chores.open())

File: family-hub-complete/client/family-hub-client.js
// Family Hub JavaScript/Node.js SDK — no npm install needed, copy this file
export class FamilyHubClient {
  constructor({ baseUrl, apiKey, timeout = 10000 }) {
    this.baseUrl=baseUrl.replace(/\/$/,''); this.apiKey=apiKey; this.timeout=timeout;
    this.events=new EventsResource(this); this.chores=new ChoresResource(this);
    this.meals=new MealsResource(this); this.grocery=new GroceryResource(this);
    this.integrations=new IntegrationsResource(this); this.system=new SystemResource(this);
  }
  async _request(method,path,body){
    const headers={'Content-Type':'application/json'};
    if(this.apiKey) headers['X-API-Key']=this.apiKey;
    const ctrl=new AbortController(); const t=setTimeout(()=>ctrl.abort(),this.timeout);
    try{
      const r=await fetch(`${this.baseUrl}${path}`,{method,headers,body:body?JSON.stringify(body):undefined,signal:ctrl.signal});
      clearTimeout(t);
      const d=await r.json().catch(()=>({}));
      if(!r.ok){const e=new Error(d.error||`HTTP ${r.status}`);e.status=r.status;throw e;}
      return d;
    }catch(e){clearTimeout(t);if(e.name==='AbortError')throw new Error('Request timed out');throw e;}
  }
  get(p){return this._request('GET',p);}
  post(p,b){return this._request('POST',p,b);}
  patch(p,b){return this._request('PATCH',p,b);}
  del(p){return this._request('DELETE',p);}
}
class EventsResource{
  constructor(c){this._c=c;}
  list(){return this._c.get('/api/events');}
  create(d){return this._c.post('/api/events',d);}
  async today(){const a=await this.list();const t=new Date().toISOString().slice(0,10);return a.filter(e=>e.start?.slice(0,10)===t).sort((a,b)=>new Date(a.start)-new Date(b.start));}
  async upcoming(days=7){const a=await this.list(),now=new Date(),then=new Date(now.getTime()+days*864e5);return a.filter(e=>{const d=new Date(e.start);return d>=now&&d<=then;}).sort((a,b)=>new Date(a.start)-new Date(b.start));}
}
class ChoresResource{
  constructor(c){this._c=c;}
  list(){return this._c.get('/api/chores');}
  create(d){return this._c.post('/api/chores',d);}
  complete(id){return this._c.post(`/api/chores/${id}/complete`);}
  async open(){return(await this.list()).filter(c=>c.status==='open');}
  async leaderboard(){const a=await this.list(),m={};for(const c of a.filter(c=>c.status==='done'))m[c.assignee]=(m[c.assignee]||0)+(c.points||1);return Object.entries(m).map(([name,points])=>({name,points})).sort((a,b)=>b.points-a.points);}
}
class MealsResource{
  constructor(c){this._c=c;}
  week(){return this._c.get('/api/meals/week');}
  create(d){return this._c.post('/api/meals',d);}
  async today(){const{meals}=await this.week();const t=new Date().toISOString().slice(0,10);return meals.filter(m=>m.date===t);}
}
class GroceryResource{
  constructor(c){this._c=c;}
  list(){return this._c.get('/api/grocery/items');}
  add(d){return this._c.post('/api/grocery/items',d);}
  addMany(items){return Promise.all(items.map(i=>this.add(i)));}
}
class IntegrationsResource{
  constructor(c){this._c=c;}
  status(){return this._c.get('/api/integrations/status');}
  google={startAuth:()=>this._c.post('/api/integrations/google/auth/start'),sync:()=>this._c.post('/api/integrations/google/sync')};
  apple={calendars:()=>this._c.get('/api/integrations/apple/calendars'),sync:()=>this._c.post('/api/integrations/apple/sync')};
  webcal={sync:()=>this._c.post('/api/integrations/webcal/sync')};
}
class SystemResource{
  constructor(c){this._c=c;}
  health(){return this._c.get('/api/health');}
  async isOnline(){try{await this.health();return true;}catch{return false;}}
}
export const createLocalClient=(ip='localhost',key)=>new FamilyHubClient({baseUrl:`http://${ip}:3001`,apiKey:key});
export const createRemoteClient=(host,key)=>new FamilyHubClient({baseUrl:`https://${host}`,apiKey:key,timeout:15000});
